### ISSUE TITLE
.NET 6 iOS support

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry-dotnet#1282]
Since .NET 6 landed (#939 ) we can look into adding `net6-ios` and including sentry-cocoa.
See: https://github.com/xamarin/xamarin-macios/blob/861c40dbe45478b848ac1d165314e05de012bf0f/tests/dotnet/NativeXCFrameworkReferencesApp/shared.csproj

When we bundle the [Sentry iOS SDK](https://github.com/getsentry/sentry-cocoa) we need to take into account that Mono uses some signals to interpret `NullReferenceException`s and need special case (run before other crash reports do):

See: https://www.mono-project.com/docs/advanced/signals/#incomplete-solution
See also: https://github.com/dotnet/runtime/issues/44736

This is probably something we can do in the binding project, before calling `Init` in the iOS SDK.

Proposed solution, based on the Mono docs:

```csharp
try {
} finally {
    Mono.Runtime.RemoveSignalHandlers ();
    try {
        // Init the Sentry iOS SDK:
        SentrySDK.start(...);
    } finally {
        Mono.Runtime.InstallSignalHandlers ();
    }
}
```

> Since Mono’s signal handlers are reinstalled and not restored, Mono will now chain to the signal handlers installed by EnableCrashReporting.

> The code is executed in a finally block, so that the Mono runtime will never abort it under any circumstance.

> It’s recommended to do this as early as possible when launching the process, in particular before starting any secondary threads. If any signals are raised between removing and reinstalling Mono’s signal handlers (this includes NullReferenceExceptions), the app will crash.

- [ ] Make sure native debug symbols are bundled in the NuGet package. See: https://github.com/xamarin/xamarin-macios/issues/9934#issuecomment-714675380

See: https://www.mono-project.com/docs/advanced/signals/#complete-solution

**Comment by gsomix** [comment]
@lucas-zimerman Hey, any news on it?

**Comment by lucas-zimerman** [comment]
> @lucas-zimerman Hey, any news on it?

Currently, it's on hold but we'll return to it near to MAUI release date.

**Comment by bruno-garcia** [comment]
@gsomix anything specific you're looking for?

The SDK should work on iOS, just not capturing native crashes

**Comment by Dolfik1** [comment]
> The SDK should work on iOS, just not capturing native crashes

This is the main problem. Native crashes are also very useful information. Many native crashes can be caused by managed code. 

It would also be important to have information about the platform, device model, os version, cpu model, etc. This is important especially on Android.

**Comment by Dolfik1** [comment]
I also want to draw attention to an exception in the  getsentry/sentry-xamarin#97

This exception is thrown inside sentry-dotnet:
https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs#L34

I don't know how this used to work, but the Xamarin.iOS implementation of HttpClient does not have an implementation for SupportsAutomaticDecompression: 
https://github.com/xamarin/xamarin-macios/blob/main/src/Foundation/NSUrlSessionHandler.cs#L599
https://github.com/xamarin/xamarin-macios/issues/13579

I checked our project (old Xamarin.iOS) and it uses `NSUrlSessionHandler` as default handler:
```
<MtouchHttpClientHandler>NSUrlSessionHandler</MtouchHttpClientHandler>
```

But it stopped working after porting the project to .Net 6 (getsentry/sentry-xamarin#97)


The recommended handler is NSUrlSessionHandler. You can read more about HTTP stack on iOS:
https://docs.microsoft.com/en-us/xamarin/cross-platform/macios/http-stack

So it seems that the problem is really in sentry-dotnet.

**Comment by bruno-garcia** [comment]
It seems the problem is really in .NET 6 in this case. In the iOS implementation, that's throwing:

```
Unhandled managed exception: Operation is not supported on this platform. 
(System.PlatformNotSupportedException)
   at System.Net.Http.NSUrlSessionHandler.get_SupportsAutomaticDecompression()
```

Perhaps disabling compression on the client could be a work around

**Comment by bruno-garcia** [comment]
The Sentry.Xamarin package relies on Xamarin.Essentials which is now part of .NET MAUI. So we could add the device information bits for MAUI (on a `Sentry.Maui`) but I don't believe we'll go through the effort of adding `#if` throughout and call into MonoTouch and Xamarin.Android APIs to get device data (basically replicating what Essentials does).

That said improving support (including a work around so that it doesn't throw when using a net6 iOS build) is needed here and we'd love to get come contributions

**Comment by bruno-garcia** [comment]
We're going to tackle this soon

**Comment by mattjohnsonpint** [comment]
I started playing with this, and and found a few things:

- With the [latest release of xamarin-macios](https://github.com/xamarin/xamarin-macios/releases/tag/dotnet-6.0.3xx), things have changed significantly.  The Xamarin SDK now uses the .NET infrastructure instead of Mono, and is now installed via .NET Workloads (`sudo dotnet workload install ios`).  It's unclear to me if that means Mono is still used at runtime or if it's completely CoreCLR now.

- From a managed-only perspective, I can't seem to catch an unhandled exception anywhere.  Some previous posts (like [this one](https://stackoverflow.com/questions/39503390/global-exception-handling-in-xamarin-cross-platform)) mention hooking `AppDomain.CurrentDomain.UnhandledException` and `TaskScheduler.UnobservedTaskException` from `FinishedLaunching` in `AppDelegate.cs`, but that doesn't seem to work now.

- When you throw an unhandled exception while debugging on Rider or Visual Studio, the IDE seems to catch the exception repeatedly until you stop debugging and then it eventually crashes the app.  Indeed, if you hook `AppDomain.CurrentDomain.FirstChanceException` and just `Console.Write` something, then throw an exception during startup and run on the command line, you'll see the event gets called over and over, flooding the screen until you eventually get a native crash.  I believe this is related to https://github.com/xamarin/xamarin-macios/issues/14796.

- There's some additional requests for a global exception handler for iOS already, such as https://github.com/xamarin/xamarin-macios/issues/12040

- The same results happen with `maccatalyst`, as would be expected.

Hopefully we can work through some of this by leveraging our native SDK.

**Comment by bruno-garcia** [comment]
Thanks for sharing these findings.

> It's unclear to me if that means Mono is still used at runtime or if it's completely CoreCLR now.

Mono is the runtime for mobile in .NET 6 👍 

> When you throw an unhandled exception while debugging on Rider or Visual Studio, the IDE seems to catch the exception repeatedly until you stop debugging and then it eventually crashes the app.

On iOS, you can't debug through a signal handler, so crash reporting is hard to debug. We test the iOS native SDK for hard crashes only without a debugger attached. Could it be related? Seems not to be the case now that I followed the issue you raised with .NET folks

Binding docs seem to still be under Xamarin only: https://docs.microsoft.com/en-us/xamarin/cross-platform/macios/binding/objective-c-libraries?tabs=macos

@filipnavara shared a link to some tests folks do on bindings that can help us out: https://github.com/xamarin/xamarin-macios/tree/f1e4cbcfcdd887819b493460441da187e0e2f11b/tests/bindings-xcframework-test/dotnet

**Comment by mattjohnsonpint** [comment]
Completed in #1829
